### PR TITLE
Fix issues with end match

### PR DESCRIPTION
**Pull request by tgodzik**
There is some logic around optional braces that relies on the match keyword, however it should not take into account `end match`. This is now fixed.

### CODE REVIEW

**Review by kitbellew — APPROVED** [review]

File: tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/EndMarkerSuite.scala [hunk 1]
```diff
@@ -151,4 +151,78 @@ class EndMarkerSuite extends BaseDottySuite {
       )
     )
   }
+
+  test("if-then-with-end") {
```
Comment (dos65) [inline tests/shared/src/test/scala/scala/meta/tests/parsers/dotty/EndMarkerSuite.scala:155]:
```suggestion
  test("match-with-end") {
```

**Review by dos65 — APPROVED** [review]

**Comment by kitbellew** [comment]
@tgodzik this one doesn't work:
```
object a:
  val (foo, boo) =
    bar
    baz
  end val
  val (foo, boo) =
    bar
    baz
  end val
```

**Comment by tgodzik** [comment]
I will take a look tomorrow!

**Comment by dos65** [comment]
@tgodzik I almost fixed it

File: scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala [hunk 2]
```diff
@@ -602,7 +604,8 @@ class ScalametaParser(input: Input)(implicit dialect: Dialect) { parser =>
                 } else
                   // always add indent for indented `match` block
                   // check the previous token to avoid infinity loop
-                  prev.is[KwMatch] && next.is[KwCase] && token.isNot[Indentation.Indent]
+                  (!prev.prev.is[soft.KwEnd] && prev.is[KwMatch]) &&
+                  next.is[KwCase] && token.isNot[Indentation.Indent]
```
Comment (kitbellew) [inline scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: sorry for being a pain but this is something that i learned the hard way in scalafmt. does prev refer to the previous token, skipping both whitespace and comments?

will this handle something like that:
```
    foo match // blah
    case bar => ...
    end match // blah
  case qux => ...
```

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: It skips them:
```scala
    def prev: Token = {
      val prev = scannerTokens.apply(Math.max(token.index - 1, 0))
      if (prev.is[Whitespace] || prev.is[Comment]) prev.prev
      else prev
    }
```

Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: what about this:
```
val end = ...
end match
case theend => ...
end match
```

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: this doesn't compile in Dotty from what I see.

Comment (kitbellew) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: why not? `end` is not restricted https://scalacenter.github.io/scala-3-migration-guide/docs/incompatibilities/syntactic-changes.html#restricted-keywords

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: Might be a bug? Or really there is a number of scenarios, where soft keywords don't work, but it's not stated.

Same with:
```scala
def extension(a : Int) = a + 2
extension(2)
```
will not work.

You need to do:
```scala
def extension(a : Int) = a + 2
`extension`(2)
```

So soft keywords are kind of dodgy and when I asked Martin he said that this is fine.

Comment (tgodzik) [reply scalameta/parsers/shared/src/main/scala/scala/meta/internal/parsers/ScalametaParser.scala:608]: https://github.com/lampepfl/dotty/issues/10076 for some context (I kind of argued with him there :sweat_smile: )

**Pull request merged**